pallets/asset-registry/src/lib.rs

```diff
@@ -617,37 +722,121 @@ impl<T: Config> GetByKey<T::AssetId, T::Balance> for Pallet<T> {
-	type Error = DispatchError;
+impl<T: Config> Inspect for Pallet<T> {
+	type AssetId = T::AssetId;
+	type Location = T::AssetNativeLocation;
 
-	fn create_asset(name: &[u8], kind: AssetKind, existential_deposit: T::Balance) -> Result<T::AssetId, Self::Error> {
-		let bounded_name: BoundedVec<u8, T::StringLimit> = Self::to_bounded_name(name.to_vec())?;
-		Pallet::<T>::register_asset(bounded_name, kind.into(), existential_deposit, None, None)
+	fn is_sufficient(id: Self::AssetId) -> bool {
+		match Self::assets(id) {
+			Some(a) => a.is_sufficient,
+			None => false,
+		}
+	}
+
+	fn exists(id: Self::AssetId) -> bool {
+		Assets::<T>::try_get(id).is_ok()
+	}
+
+	fn decimals(id: Self::AssetId) -> Option<u8> {
+		Self::assets(id).and_then(|a| a.decimals)
+	}
+
+	fn asset_type(id: Self::AssetId) -> Option<AssetKind> {
+		Self::assets(id).map(|a| a.asset_type.into())
 	}
-}
 
-impl<T: Config> InspectRegistry<T::AssetId> for Pallet<T> {
-	fn exists(asset_id: T::AssetId) -> bool {
-		Assets::<T>::contains_key(asset_id)
+	fn is_blacklisted(id: Self::AssetId) -> bool {
+		BlacklistedAssets::<T>::contains_key(id)
 	}
 
-	fn decimals(asset_id: T::AssetId) -> Option<u8> {
-		Some(AssetMetadataMap::<T>::get(asset_id)?.decimals)
+	fn asset_name(id: Self::AssetId) -> Option<Vec<u8>> {
+		Self::assets(id).and_then(|a| a.name.map(|v| v.into()))
+	}
+
+	fn asset_symbol(id: Self::AssetId) -> Option<Vec<u8>> {
+		Self::assets(id).and_then(|a| a.symbol.map(|v| v.into()))
+	}
+}
+
+impl<T: Config> Mutate for Pallet<T> {
+	type Error = DispatchError;
+
+	fn set_location(asset_id: Self::AssetId, location: T::AssetNativeLocation) -> Result<(), Self::Error> {
+		ensure!(Self::exists(asset_id), Error::<T>::AssetNotFound);
+
+		Self::do_set_location(asset_id, location)
 	}
+}
+
+impl<T: Config> Create<Balance> for Pallet<T> {
+	type Error = DispatchError;
 
-	fn asset_name(asset_id: T::AssetId) -> Option<Vec<u8>> {
-		let asset = Assets::<T>::get(asset_id)?;
-		Some(asset.name.into_inner())
+	fn register_asset(
+		asset_id: Option<Self::AssetId>,
+		name: Option<&[u8]>,
+		kind: AssetKind,
+		existential_deposit: Option<Balance>,
+		symbol: Option<&[u8]>,
+		decimals: Option<u8>,
+		location: Option<Self::Location>,
+		xcm_rate_limit: Option<Balance>,
+		is_sufficient: bool,
+	) -> Result<Self::AssetId, Self::Error> {
+		let bounded_name = Self::try_into_bounded(name.map(|x| x.to_vec()))?;
+		let bounded_symbol = Self::try_into_bounded(symbol.map(|x| x.to_vec()))?;
+
+		let details = AssetDetails::new(
+			bounded_name,
+			kind.into(),
+			existential_deposit.unwrap_or(DEFAULT_ED),
+			bounded_symbol,
+			decimals,
+			xcm_rate_limit,
+			is_sufficient,
+		);
+
+		Self::do_register_asset(asset_id, &details, location)
 	}
 
-	fn asset_symbol(asset_id: T::AssetId) -> Option<Vec<u8>> {
-		let asset_metadata = AssetMetadataMap::<T>::get(asset_id)?;
-		Some(asset_metadata.symbol.into_inner())
+	fn get_or_register_asset(
```

Comment:
hm..there is another function get_or_create_asset. It does basically the same. 

i wonder if it cannot be used - one way or another,.

Reply:
I missed this one, removed